File: solutions/_1189_maximum_number_of_balloons/solution.py
from collections import Counter


class Solution:
    def maxNumberOfBalloons(self, text: str) -> int:
        countText = Counter(text)
        balloon = Counter("balloon")

        res = len(text)
        for c in balloon:
            res = min(res, countText[c] // balloon[c])
        return res


# https://leetcode.com/problems/maximum-number-of-balloons/

File: solutions/_1189_maximum_number_of_balloons/solution.java
package _1189_maximum_number_of_balloons;

import java.util.HashMap;

class Solution {

  public int maxNumberOfBalloons(String text) {
    Counter inputCtr = new Counter(text);
    Counter targetCtr = new Counter("balloon");

    int res = Integer.MAX_VALUE;
    for (char c : "balloon".toCharArray()) {
      int count = inputCtr.get(c) / targetCtr.get(c);
      res = Math.min(res, count);
    }

    return res;
  }

  public class Counter {

    private HashMap<Character, Integer> _ctr = new HashMap<>();

    public Counter(String s) {
      for (char c : s.toCharArray()) {
        _ctr.put(c, _ctr.getOrDefault(c, 0) + 1);
      }
    }

    public int get(char c) {
      return _ctr.getOrDefault(c, 0);
    }
  }
}
/*
https://leetcode.com/problems/maximum-number-of-balloons/
*/
